File: easy/141-linked-list-cycle/go/solution.go
package linked_list_cycle

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	if head == nil {
		return false
	}

	if head.Next == nil {
		return false
	}

	fast, slow := 0, 0
	fastNode := head
	slowNode := head

	for fastNode != nil && slowNode != nil {
		fast += 2
		slow += 1
		fastNode = fastNode.Next
		if fastNode == nil || fastNode.Next == nil {
			return false
		} else {
			fastNode = fastNode.Next
		}
		slowNode = slowNode.Next

		if fastNode == slowNode {
			return true
		}
	}

	return false
}

File: easy/141-linked-list-cycle/python/solution.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head
        index = 0
        while fast and fast.next:
            if fast == slow:
                if index > 0:
                    return True
                else:
                    index = 1
            slow = slow.next
            fast = fast.next.next
        return False
